<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

In [2]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
# данные удалены после выполнения заданий

In [3]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
# текст запроса
query_3_1 = f'''
select count(id)
from vacancies v
'''

In [5]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\gtsap\AppData\Local\Temp\ipykernel_16328\1114379196.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [6]:
# текст запроса
query_3_2 = f'''
select count(id)
from employers e
'''

In [7]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\gtsap\AppData\Local\Temp\ipykernel_16328\3062646138.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
# текст запроса
query_3_3 = f'''
select count(id)
from areas a
'''

In [9]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\gtsap\AppData\Local\Temp\ipykernel_16328\928149064.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
# текст запроса
query_3_4 = f'''
select count(id)
from industries i
'''

In [11]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\gtsap\AppData\Local\Temp\ipykernel_16328\2101691108.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

***

# выводы по предварительному анализу данных
Предварительный анализ данных позволил нам посмотреть структуру таблиц, получить представление о количестве вакансий (49197), содержащихся в базе, количестве работадателей (23501), количеству регионов (1362) и количеству направлений деятельности (294).

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [12]:
# текст запроса
query_3_5 = f'''
select a.name, count(area_id)
from vacancies v
left join areas a on v.area_id = a.id
group by a.name
order by 2 desc
'''

In [13]:
# результат запроса
df = pd.read_sql_query(query_3_5, connection)
df.head(5)


C:\Users\gtsap\AppData\Local\Temp\ipykernel_16328\3857530017.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_5, connection)


name  count
0           Москва   5333
1  Санкт-Петербург   2851
2            Минск   2112
3      Новосибирск   2006
4           Алматы   1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [14]:
# текст запроса
query_3_6 = f'''
select count(id)
from vacancies v
where (salary_from > 0 or salary_to > 0)
'''

In [15]:
# результат запроса
df = pd.read_sql_query(query_3_6, connection)
df

C:\Users\gtsap\AppData\Local\Temp\ipykernel_16328\4138296559.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_6, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [16]:
# текст запроса
query_3_7 = f'''
select avg(salary_from), avg(salary_to)
from vacancies v
'''

In [17]:
# результат запроса
df = pd.read_sql_query(query_3_7, connection)
df

C:\Users\gtsap\AppData\Local\Temp\ipykernel_16328\1920805602.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_7, connection)


avg            avg
0  71064.657901  110536.741923

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [18]:
# текст запроса
query_3_8 = f'''
select schedule, employment, count(id)
from vacancies v
group by schedule, employment
order by 3 desc
limit(5)
'''

In [19]:
# результат запроса
df = pd.read_sql_query(query_3_8, connection)
df

C:\Users\gtsap\AppData\Local\Temp\ipykernel_16328\2476473451.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_8, connection)


schedule           employment  count
0       Полный день     Полная занятость  35367
1  Удаленная работа     Полная занятость   7802
2     Гибкий график     Полная занятость   1593
3  Удаленная работа  Частичная занятость   1312
4    Сменный график     Полная занятость    940

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [20]:
# текст запроса
query_3_9 = f'''
select experience, count(id)
from vacancies v
group by experience
order by 2 
limit(5)
'''

In [21]:
# результат запроса
df = pd.read_sql_query(query_3_9, connection)
df

C:\Users\gtsap\AppData\Local\Temp\ipykernel_16328\3592149368.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_9, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

# выводы по детальному анализу вакансий
Детальный анализ вакансий позволил нам более подробно изучить структуру вакансий в различных разрезах. В частности мы определили количество вакансий по регионам, количество вакансий у которых заполнено хотя бы одно из занчений зарплатной вилки, определили среднее значение верхней и нижней границы зарплатной вилки, разобрались в предпочтениях по рабочему графику и типу занятости, а также структурировали вакансии по требуемому опыту работы.
Лидерами по количеству вакансий являются крупные города: Москва, Санкт-Петербург, Минск, Новосибирск, Алматы. Набольшее количество вакансий (35 367) связано с Полным рабочим днем, а также наиболее распространенный диапазон опыта это от 1 года до 3 лет.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [22]:
# текст запроса
query_3_10 = f'''
select e.name, count(*)
from vacancies v
left join employers e on v.employer_id = e.id
group by e.name
order by 2 desc
limit(5)
'''

In [23]:
# результат запроса
df = pd.read_sql_query(query_3_10, connection)
df

C:\Users\gtsap\AppData\Local\Temp\ipykernel_16328\2630767671.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_10, connection)


name  count
0         Яндекс   1933
1     Ростелеком    491
2       Тинькофф    444
3           СБЕР    428
4  Газпром нефть    331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [24]:
# текст запроса
query_3_11 = f'''
select a.name, count(v.id) count_vac, count(e.id) count_emp
from areas a
left join vacancies v on a.id = v.area_id
left join employers e on a.id = e.area
group by a.name
having count(v.id) = 0
order by 3 desc
limit(5)
'''

In [25]:
# результат запроса
df = pd.read_sql_query(query_3_11, connection)
df

C:\Users\gtsap\AppData\Local\Temp\ipykernel_16328\3267443505.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_11, connection)


name  count_vac  count_emp
0              Россия          0        410
1           Казахстан          0        207
2  Московская область          0         75
3  Краснодарский край          0         19
4            Беларусь          0         18

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [26]:
# текст запроса
query_3_12 = f'''
select e.name, count(distinct area_id)
from vacancies v
left join employers e on v.employer_id = e.id
group by e.name
order by 2 desc
limit(5)
'''

In [27]:
# результат запроса
df = pd.read_sql_query(query_3_12, connection)
df

C:\Users\gtsap\AppData\Local\Temp\ipykernel_16328\4223621131.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_12, connection)


name  count
0                  Яндекс    181
1              Ростелеком    152
2              Спецремонт    116
3  Поляков Денис Иванович     88
4               ООО ЕФИН      71

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [28]:
# текст запроса
query_3_13 = f''' 
select name, count(industry_id)
from employers e 
left join employers_industries ei on e.id = ei.employer_id
group by id
having count(industry_id) = 0
'''

In [29]:
# результат запроса
df = pd.read_sql_query(query_3_13, connection)
df

C:\Users\gtsap\AppData\Local\Temp\ipykernel_16328\1465250533.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_13, connection)


name  count
0     Северо-Западный Информационно-Мониторинговый Ц...      0
1                                АВГ Инженерные Системы      0
2                                               Лидстех      0
3                                   Евроклимат Тольятти      0
4                               Finstar Financial Group      0
...                                                 ...    ...
8414                            ОГКУ ЦЗН города Бодайбо      0
8415                               ПП «Южуралкомсервис»      0
8416                   Whitecliff Investment Management      0
8417                                       TransPerfect      0
8418                                 GLOSS|Студия Кухни      0

[8419 rows x 2 columns]

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [30]:
# текст запроса
query_3_14 = f''' 
select name, count(industry_id)
from employers e 
left join employers_industries ei on e.id = ei.employer_id
group by id
having count(industry_id) = 4
order by 1
'''

In [31]:
# результат запроса
df = pd.read_sql_query(query_3_14, connection)
df

C:\Users\gtsap\AppData\Local\Temp\ipykernel_16328\4196720133.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_14, connection)


name  count
0        101 Интернет      4
1            21vek.by      4
2                2ГИС      4
3                  2К      4
4         4 пикселя +      4
...               ...    ...
1133            ЮРИОН      4
1134  ЮТИП Технологии      4
1135           ЯКласс      4
1136            ЯрНео      4
1137    Яшин&Партнёры      4

[1138 rows x 2 columns]

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [32]:
# текст запроса
query_3_15 = f''' 
select count(*)
from employers_industries ei
join industries i on ei.industry_id = i.id
where name = 'Разработка программного обеспечения'
'''

In [33]:
# результат запроса
df = pd.read_sql_query(query_3_15, connection)
df

C:\Users\gtsap\AppData\Local\Temp\ipykernel_16328\1079631588.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_15, connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [34]:
# код для получения списка городов-милионников
ds = pd.read_html("https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8")
ds

[    0                                                  1
 0 NaN  Информация в этой статье или некоторых её разд...,
      №            Город Население, тыс. чел.  (на 1 января 2023 года)  \
 0    1           Москва                                        13 098   
 1    2  Санкт-Петербург                                         5 598   
 2    3      Новосибирск                                         1 635   
 3    4     Екатеринбург                                         1 539   
 4    5           Казань                                         1 315   
 5    6  Нижний Новгород                                           NaN   
 6    7       Красноярск                                         1 197   
 7    8        Челябинск                                         1 183   
 8    9           Самара                                         1 164   
 9   10              Уфа                                         1 158   
 10  11   Ростов-на-Дону                                           Na

In [35]:
query_3_166 = f''' 
select *
from areas 
where name = 'Нижний Новгород'
'''
df = pd.read_sql_query(query_3_166, connection)
df

C:\Users\gtsap\AppData\Local\Temp\ipykernel_16328\1651353236.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_166, connection)


id             name
0  66  Нижний Новгород

In [36]:
# текст запроса
query_3_16 = f''' 
select a.name, count(*) cnt
from vacancies v
left join areas a on v.area_id = a.id
left join employers e on v.employer_id = e.id
group by a.name, e.name
having (e.name = 'Яндекс' and (a.name = 'Москва' or a.name = 'Санкт-Петербург' or a.name = 'Новосибирск' or a.name = 'Екатеринбург' or a.name = 'Казань' or a.name = 'Нижний Новгород' or a.name = 'Красноярск' or a.name = 'Челябинск' or a.name = 'Самара' or a.name = 'Уфа' or a.name = 'Ростов-на-Дону' or a.name = 'Омск' or a.name = 'Краснодар' or a.name = 'Воронеж' or a.name = 'Пермь' or a.name = 'Волгоград'))
order by 2 desc

'''

In [37]:
# результат запроса
df = pd.read_sql_query(query_3_16, connection)
df

C:\Users\gtsap\AppData\Local\Temp\ipykernel_16328\3189269934.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_16, connection)


name  cnt
0            Москва   54
1   Санкт-Петербург   42
2      Екатеринбург   39
3   Нижний Новгород   36
4       Новосибирск   35
5           Воронеж   32
6         Краснодар   30
7            Самара   26
8               Уфа   26
9            Казань   25
10            Пермь   25
11   Ростов-на-Дону   25
12        Волгоград   24
13       Красноярск   23
14        Челябинск   22
15             Омск   21

***

# выводы по анализу работодателей
В ходе детального работодателей мы посмотрели рейтинг работодателей по количеству вакансий, определили для каждого работодателя количество регионов в которых он присутствует и число вакансий в каждом регионе, определили регионы в которых нет вакансий, посмотрели работодателей в связке со сферой деятельности, вывели количество вакансий в разрезе городов-миллионников для конкретного работодателя.


# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [100]:
# текст запроса
query_3_17 = f''' 
select count(id)
from vacancies v
where lower(name) like'%data%'
or lower(name) like'%данн%'

'''

In [101]:
# результат запроса
df = pd.read_sql_query(query_3_17, connection)
df

C:\Users\gtsap\AppData\Local\Temp\ipykernel_16328\3839030628.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_17, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [255]:
# текст запроса
query_3_18 = f''' 
select *
from vacancies v
where (lower(name) like'%data scientist%'
or lower(name) like'%data science%'
or lower(name) like'%исследователь данных%'
or lower(name) like'%machine learning%'
or lower(name) like'%машинн%обучен%'
or name like'%ML%')
and (lower(name) like'%junior%'
or experience = 'Нет опыта'
or schedule = 'Стажировка')
except
select *
from vacancies v
where name like'%HTML%'
'''

In [256]:
# результат запроса
df = pd.read_sql_query(query_3_18, connection)
df

C:\Users\gtsap\AppData\Local\Temp\ipykernel_16328\490100124.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_18, connection)


id                                               name  \
0   54945920              Python developer / Junior ML engineer   
1   66142163  Junior Data Scientist (математик-аналитик) уда...   
2   55373640  Junior Data Scientist Управления модельных рис...   
3   50996908          Senior Research Engineer / Data Scientist   
4   55529384     Математик, Junior/Middle/Senior Data Scientist   
5   55127396  Junior Data Scientist (Валидация скоринговых м...   
6   55104548       Риск-аналитик (ML Engineer / Data Scientist)   
7   55127397  Junior Data Scientist (Валидация скоринговых м...   
8   54924841  Стажер направления Машинное обучение и анализ ...   
9   55127392  Junior Data Scientist (Валидация скоринговых м...   
10  55145641                              Junior Data Scientist   
11  55127391  Junior Data Scientist (Валидация скоринговых м...   
12  55046966             Junior Data Scientist for NLP (Python)   
13  55343711  Data Scientist/Machine Learning Engineer (Deve...   
14  55127398  Junior Data Scientist (Валидация скоринговых м...   
15  54830582                              Junior Data Scientist   
16  55120211                         Junior Data Scientist (VK)   
17  54522206            Data Scientist/ML engineer (Python/C++)   
18  55552998  Математик-аналитик (data scientist / data anal...   
19  54702840  Программист Python/ Data Scientist (junior / m...   
20  54825318                           Data Scientist (Junior+)   
21  55156731  Стажёр Data Scientist / Группа сервисов Антифр...   
22  54948266                         Senior Data Scientist (AI)   
23  55127395  Junior Data Scientist (Валидация скоринговых м...   
24  50072394                 Research Engineer / Data Scientist   
25  55127394  Junior Data Scientist (Валидация скоринговых м...   
26  54657714                    Data Scientist (junior+/middle)   
27  55127393  Junior Data Scientist (Валидация скоринговых м...   
28  55343710  Data Scientist/Machine Learning Engineer (Deve...   
29  55553897                     Junior Data Scientist (Стажер)   
30  54475492                            Data Scientist (Junior)   
31  52396404                             Data Scientist, Junior   
32  52158791  Исследователь в области машинного обучения и м...   
33  55204943                              Ведущий AI/ML инженер   
34  55120401                       Junior/Middle Data Scientist   
35  55095363            Риск - Аналитик (Junior Data scientist)   
36  55074464        Data scientist (Junior/Middle) в консалтинг   
37  50533416                            Junior+ CV/ML developer   
38  55127390  Junior Data Scientist (Валидация скоринговых м...   
39  54556346                              Ведущий AI/ML инженер   
40  54949463          Junior Research Engineer / Data Scientist   
41  55895592             Junior Data Scientist (Promo Forecast)   
42  55052496                              Junior Data Scientist   
43  55236061          Data Scientist (Computer Vision) / Junior   
44  54689300                                     Data Scientist   
45  55390769          Data Scientist (Computer Vision) - Junior   
46  55119989                       Intern/Junior Data Scientist   
47  54882705       Junior Data scientist/ Младший риск-аналитик   
48  55343709  Data Scientist/Machine Learning Engineer (Deve...   
49  55064629                              Data Scientist в Сбер   
50  52396406                             Data Scientist, Junior   

                                           key_skills          schedule  \
0   Python\tC++\tОптимизация кода\tАнализ кода\tМа...     Гибкий график   
1   SQL\tMS SQL\tУдаленная работа\tPython\tМатемат...  Удаленная работа   
2   Математическое моделирование\tАнализ данных\tС...       Полный день   
3   PyTorch\tMachine Learning\tКонференции\tPython...       Полный день   
4   Python\tMATLAB\tМатематическое моделирование\t...  Удаленная работа   
5   Оценка рисков\tMS PowerPoint\tСтатистический а...       Полный день   


3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [249]:
# текст запроса
query_3_19 = f''' 
select *
from vacancies v
where (lower(name) like'%data scientist%'
or lower(name) like'%data science%'
or lower(name) like'%исследователь данных%'
or lower(name) like'%machine learning%'
or lower(name) like'%машинн%обучен%'
or name like'%ML%')
and (key_skills like'%SQL%' or key_skills like'%Postgres%')
except
select *
from vacancies v
where name like'%HTML%'
'''

In [250]:
# результат запроса
df = pd.read_sql_query(query_3_19, connection)
df

C:\Users\gtsap\AppData\Local\Temp\ipykernel_16328\2579909047.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_19, connection)


id                                            name  \
0    54716166           Руководитель направления Data Science   
1    55120847                                  Data Scientist   
2    54854359                                  Data scientist   
3    55127484  Data Scientist (Валидация скоринговых моделей)   
4    54825677                               Senior ML-инженер   
..        ...                                             ...   
196  54545779                            Data Science эксперт   
197  54618365                      Senior Data Scientist (UA)   
198  55042690                                  Разработчик ML   
199  54882985                 ML Research Engineer (MY.GAMES)   
200  54659098                 Data Scientist / ML-разработчик   

                                            key_skills          schedule  \
0    Python\tМатематическая статистика\tSQL\tLinux\...       Полный день   
1    Python\tSQL\tData Mining\tАналитические исслед...       Полный день   
2    Python\tSQL\tHadoop\tSpark\tBig Data\tМашинное...       Полный день   
3    Оценка рисков\tMS PowerPoint\tСтатистический а...       Полный день   
4    Python\tSQL\tHadoop\tJava\tAdobe Photoshop\tSp...  Удаленная работа   
..                                                 ...               ...   
196  Python\tСтатистический анализ\tSQL\tМатематиче...       Полный день   
197    Python\tSQL\tTableau\tМатематическая статистика  Удаленная работа   
198                   Python\tGit\tSQL\tLinux\tPyTorch       Полный день   
199  Python\tA/B тесты\tМатематическая статистика\tSQL       Полный день   
200                  Python\tSQL\tML\tMachine Learning       Полный день   

             experience        employment  salary_from  salary_to  area_id  \
0         От 3 до 6 лет  Полная занятость          NaN   300000.0        1   
1    От 1 года до 3 лет  Полная занятость          NaN        NaN        2   
2         От 3 до 6 лет  Полная занятость          NaN        NaN        1   
3    От 1 года до 3 лет  Полная занятость          NaN        NaN       66   
4           Более 6 лет  Полная занятость          NaN        NaN        1   
..                  ...               ...          ...        ...      ...   
196  От 1 года до 3 лет  Полная занятость          NaN        NaN       88   
197  От 1 года до 3 лет  Полная занятость          NaN        NaN        2   
198  От 1 года до 3 лет  Полная занятость          NaN        NaN        1   
199  От 1 года до 3 лет  Полная занятость          NaN        NaN        1   
200  От 1 года до 3 лет  Полная занятость          NaN        NaN        1   

     employer_id  
0          24209  
1           2575  
2           3127  
3           4181  
4        1429999  
..           ...  
196        47939  
197       114448  
198        24947  
199        15478  
200      2942396  

[201 rows x 10 columns]

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [257]:
# текст запроса
query_3_20 = f''' 
select *
from vacancies v
where (lower(name) like'%data scientist%'
or lower(name) like'%data science%'
or lower(name) like'%исследователь данных%'
or lower(name) like'%machine learning%'
or lower(name) like'%машинн%обучен%'
or name like'%ML%')
and (key_skills like'%Python%')
except
select *
from vacancies v
where name like'%HTML%'
'''

In [258]:
# результат запроса
df = pd.read_sql_query(query_3_20, connection)
df

C:\Users\gtsap\AppData\Local\Temp\ipykernel_16328\855317856.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_20, connection)


id                                           name  \
0    54689300                                 Data Scientist   
1    55098469         Data Scientist в Data Accelerator (ML)   
2    54618365                     Senior Data Scientist (UA)   
3    54662354  ML Engineer в области рекомендательных систем   
4    54788880                          Senior Data Scientist   
..        ...                                            ...   
346  55004712               Middle MLE (Big Data Department)   
347  66120847                          AI/ML Senior Engineer   
348  54753402                                 Data Scientist   
349  54925403                                 Data Scientist   
350  55204390                          Middle Data Scientist   

                                            key_skills          schedule  \
0    Python\tGit\tTensorFlow\tKeras\tJupiter\tNLP\t...       Полный день   
1                  Python\tSQL\tGit\tАнализ данных\tML     Гибкий график   
2      Python\tSQL\tTableau\tМатематическая статистика  Удаленная работа   
3    ML\tDL\tBig Data\tPython\tSCALA\tMachine Learning       Полный день   
4    Python\tMachine Learning\tData Analysis\tNumpy...  Удаленная работа   
..                                                 ...               ...   
346                 Python\tGit\tSQL\tScrum\tML\tSpark       Полный день   
347                     Python\tFlask\tLinux\tAWS\tGit     Гибкий график   
348  Python\tМатематическое моделирование\tМатемати...       Полный день   
349       Python\tData Science\tUnix\tDocker\tRabbitMQ  Удаленная работа   
350        Python\tLearning to Rank\tData Science\tNLP  Удаленная работа   

             experience        employment  salary_from  salary_to  area_id  \
0             Нет опыта  Полная занятость          NaN        NaN        4   
1         От 3 до 6 лет  Полная занятость          NaN        NaN        1   
2    От 1 года до 3 лет  Полная занятость          NaN        NaN        2   
3         От 3 до 6 лет  Полная занятость          NaN        NaN        1   
4           Более 6 лет  Полная занятость          NaN        NaN       74   
..                  ...               ...          ...        ...      ...   
346  От 1 года до 3 лет  Полная занятость          NaN        NaN        1   
347       От 3 до 6 лет  Полная занятость          NaN        NaN     2759   
348  От 1 года до 3 лет  Полная занятость     250000.0   300000.0        1   
349  От 1 года до 3 лет  Полная занятость          NaN        NaN        1   
350       От 3 до 6 лет  Полная занятость          NaN        NaN       66   

     employer_id  
0           2791  
1           8582  
2         114448  
3           3529  
4           6189  
..           ...  
346         7172  
347       204511  
348      2613123  
349         6528  
350        41862  

[351 rows x 10 columns]

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [361]:
# текст запроса
query_3_21 = f''' 
select *, (length(key_skills) - length(replace(key_skills, CHR (9), ''))+1) as len
from vacancies v
where (lower(name) like'%data scientist%'
or lower(name) like'%data science%'
or lower(name) like'%исследователь данных%'
or lower(name) like'%machine learning%'
or lower(name) like'%машинн%обучен%'
or name like'%ML%')
except
select *, (length(key_skills) - length(replace(key_skills, CHR (9), ''))+1) as len
from vacancies v
where name like'%HTML%'
'''

In [363]:
# результат запроса
df = pd.read_sql_query(query_3_21, connection)
df['len'].mean()

C:\Users\gtsap\AppData\Local\Temp\ipykernel_16328\2458972748.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_21, connection)


6.406032482598608

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [405]:
# текст запроса
query_3_22 = f''' 
select *, (COALESCE(salary_from, salary_to) + COALESCE(salary_to, salary_from)) / 2 as salary
from vacancies v
where (lower(name) like'%data scientist%'
or lower(name) like'%data science%'
or lower(name) like'%исследователь данных%'
or lower(name) like'%machine learning%'
or lower(name) like'%машинн%обучен%'
or name like'%ML%')
and (salary_from is not null or salary_to is not null)
except
select *, (COALESCE(salary_from, salary_to) + COALESCE(salary_to, salary_from)) / 2 as salary
from vacancies v
where name like'%HTML%'
'''

In [413]:
# результат запроса
df = pd.read_sql_query(query_3_22, connection)
df.groupby(['experience']).mean('salary')

C:\Users\gtsap\AppData\Local\Temp\ipykernel_16328\3699072493.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_22, connection)


id    salary_from      salary_to     area_id  \
experience                                                                   
Нет опыта           5.468939e+07   61000.000000   97000.000000    5.142857   
От 1 года до 3 лет  5.489643e+07  121272.000000  175672.800000   99.678571   
От 3 до 6 лет       5.521207e+07  198948.631579  290589.368421  219.111111   

                     employer_id         salary  
experience                                       
Нет опыта           2.282906e+06   74642.857143  
От 1 года до 3 лет  2.738415e+06  139674.750000  
От 3 до 6 лет       3.029660e+06  243114.666667

***

# выводы по предметному анализу
В ходе предметного анализа мы отфильтровали данные относящиеся к DS и посмотрели различные комбинации, такие как количество вакансий для начинающих DS, насколько популярны навыки SQL b Python в отобранных вакансиях, среднее количество запрашиваемых навыков для вакансий, а также средние зарплаты по категориям опыта соискателей.
Среднее количество навыков составляет 6,41. Средняя зарплата увеличивается по мере роста опыта, так при его отсутствии для DS она составляет 74,6 т.р., от года до 3 лет уже порядка 140 т.р., а для диапазона от 3-х лет до 6-ти ужу 243 т.р. - увеличение почти что в два раза для каждой группы опыта.

# Общий вывод по проекту

# подведем итог исследования, обобщите выводы
В целом если рассмтеривать количество вакансий для DS, то их доля в общем количестве вакансий достаточно мало: 480 / 49197 = 0,1%, в тоже время в данной отрасли имеется неплохие перспективы по заработной плате, которая практически удваивается за каждые три года (без опыта - 74 000, от 1 года до 3 лет - 139 000, от 3 лет до 6 лет - 243 000). Однако и требования достаточно весомы: в среднем надо иметь более 6 ключевых навыков по профессии, знать SQL и Python, эти навыки востребованы в большинстве вакансий DS. Кроме того, в целом большинство вакансий сосредоточено в крупных городах, а наиболее крупные работодатели связаны с ИТ отраслью, банковской, нефтяной отраслью и телекоммуникациями.